crates/runtime/src/traphandlers.rs

```diff
@@ -397,3 +429,120 @@ mod tls {
+/// that is large enough and has a guard page.
+#[cfg(unix)]
+mod unix_sigaltstack {
+    use std::cell::RefCell;
+    use std::convert::TryInto;
+    use std::ptr::null_mut;
+
+    thread_local! {
+        /// Thread-local state is lazy-initialized on the first time it's used, and dropped
+        /// when the thread exits.
+        pub(crate) static TLS: RefCell<Tls> = RefCell::new(unsafe { Tls::new() });
+    }
+
+    /// The size of the sigaltstack (not including the guard, which will be added).
+    /// Make this large enough to run our signal handlers.
+    static STACK_SIZE: usize = libc::SIGSTKSZ * 4;
+
+    /// A utility to register a new sigaltstack.
+    pub(crate) struct Tls {
+        guard_size: usize,
+        sigaltstack_alloc_size: usize,
+        new_stack: libc::stack_t,
+    }
+
+    impl Tls {
+        unsafe fn new() -> Self {
+            let page_size: usize = libc::sysconf(libc::_SC_PAGESIZE).try_into().unwrap();
+            let guard_size = page_size;
+
+            let mut me = Self {
+                guard_size,
+                sigaltstack_alloc_size: guard_size + STACK_SIZE,
+                new_stack: libc::stack_t {
+                    ss_sp: null_mut(),
+                    ss_flags: 0,
+                    ss_size: 0,
+                },
+            };
+
+            // Allocate memory.
+            let ptr = libc::mmap(
+                null_mut(),
+                me.sigaltstack_alloc_size,
+                libc::PROT_NONE,
+                libc::MAP_PRIVATE | libc::MAP_ANON,
+                -1,
+                0,
+            );
+            if ptr == libc::MAP_FAILED {
+                super::raise_oom_trap()
+            }
+
+            // Prepare the stack, register it, and sanity check the old stack.
+            let stack_ptr = (ptr as usize + me.guard_size) as *mut libc::c_void;
+            me.new_stack = libc::stack_t {
+                ss_sp: stack_ptr,
+                ss_flags: 0,
+                ss_size: STACK_SIZE,
+            };
+            let r = libc::mprotect(stack_ptr, STACK_SIZE, libc::PROT_READ | libc::PROT_WRITE);
+            assert_eq!(r, 0, "mprotect to configure memory for sigaltstack failed");
+
+            // Register the new alternate stack (and record the old one, so we can sanity-check it).
+            let mut old_stack = libc::stack_t {
+                ss_sp: null_mut(),
+                ss_flags: 0,
+                ss_size: 0,
+            };
+            let r = libc::sigaltstack(&mut me.new_stack, &mut old_stack);
+            assert_eq!(r, 0, "registering new sigaltstack failed");
+
+            // Sanity-check the old stack.
+            assert_eq!(
+                old_stack.ss_flags, 0,
+                "old sigaltstack had unexpected flags set"
+            );
+            assert!(
+                old_stack.ss_size <= STACK_SIZE,
+                "old sigaltstack had a larger stack"
+            );
```

Comment:
I think these checks may want to move up to the front actually and not be assertions but rather runtime checks. With the C embedding we're not actually guaranteed the sigaltstack is set up (which I'm just now realizing is probably bad) because the Rust main function didn't happen. Additionally the application may have its own sigaltstack handled somewhere else.

I think we'll want to check the current sigaltstack here and if it's big enough just skip all this entirely, and then for flags I'm not really sure we need to check too much, but I don't think we should panic if the embedder has a particular configuration.